### ISSUE TITLE
API for all contrib projects

### ISSUE THREAD

**Issue by stpaultim** [backdrop-ops/backdropcms.org#870]
Do we already provide any way for third parties to access information on all our contrib projects, other than scraping the data from the site? Should we?

Or would someone really have to pull the data from Github, which might be a better and suitable alternative. I'm not sure if we have some information available or structured in a more useful format on BackdropCMS.org than it would be on Github.

**Comment by BWPanda** [comment]
Since BackdropCMS.org gets info from GitHub, I'd suggest others do the same (saves us having to create our own API for the same thing).

It seems quite simple, for example: https://api.github.com/orgs/backdrop-contrib/repos
See https://docs.github.com/en/rest/reference/repos#list-organization-repositories

**Comment by stpaultim** [comment]
Well, the difference would be that BackdropCMS.org will have information that the Github API will not have. For example information in the .info file can be processed by BackdropCMS.org. 

While that information is available in Github, I'm assuming it is a little more difficult to access.

**Comment by stpaultim** [comment]
I would like to experiment a bit with a website that lists all Backdrop contrib projects and allows us to sort and filter by category, contributor, dependencies, colors (themes), etc. 

This is information that we have or might be including in the .info file in the future. I think it will be much harder to get this directly from Github than it would be if it were available on BackdropCMS.org. 

But, I could be wrong about this, which is why I'm asking. I assume that to get this from Github, I'd have to parse the .info file myself.

**Comment by BWPanda** [comment]
> I assume that to get this from Github, I'd have to parse the .info file myself.

Yes, but there's a function for that: https://docs.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_parse_info_file/1

**Comment by BWPanda** [comment]
Here's an example I just whipped up:

```php
// Array to store GitHub/.info file data.
$data = array();

// Get information about Backdrop Contrib repos (in this case, the first 3 results).
$ch = curl_init('https://api.github.com/orgs/backdrop-contrib/repos?per_page=3&page=1');
curl_setopt($ch, CURLOPT_HTTPHEADER, array(
  'User-Agent: BackdropCMS Contrib',
  'Accept: application/vnd.github.v3+json',
));
curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
$response = curl_exec($ch);
curl_close($ch);
$repos = json_decode($response, TRUE);

foreach ($repos as $repo) {
  // Store some data from GitHub.
  $data[$repo['name']] = array(
    'description' => $repo['description'],
    'stars' => $repo['stargazers_count'],
    'forks' => $repo['forks_count'],
    'default_branch' => $repo['default_branch'],
  );

  // Get information about each project/repo from its .info file.
  $ch = curl_init('https://api.github.com/repos/backdrop-contrib/' . $repo['name'] .'/contents/' . $repo['name'] . '.info');
  curl_setopt($ch, CURLOPT_HTTPHEADER, array(
    'User-Agent: BackdropCMS Contrib',
    'Accept: application/vnd.github.VERSION.raw',
  ));
  curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
  $response = curl_exec($ch);
  curl_close($ch);
  $info = backdrop_parse_info_format($response);

  // Store some data from the .info file.
  $data[$repo['name']]['package'] = $info['package'];
  $data[$repo['name']]['configure'] = $info['configure'];
}

// Display collected data.
dpm($data);
```

It produces this:

![image](https://user-images.githubusercontent.com/2385329/153861441-b2b442ac-3393-4f83-8314-b1a8c5719dbd.png)

That's a combination of GitHub info and `.info` file info. Feel free to ~steal~ use it if it's helpful. The code could use some cleaning up/optimising, but you get the idea.

**Comment by BWPanda** [comment]
@stpaultim Has this question been answered now?

**Comment by docwilmot** [comment]
We should do this formally. Drupal has [this resolved issue](https://www.drupal.org/project/project/issues/669910) which established a regular dump of all projects to a file. Currently in our core code to get a list of all Backdrop projects in Installer module, [we do this madness here](https://github.com/backdrop/backdrop/blob/57bc2e7b22a08a412a5a44e07b7a2fc21fec28e2/core/modules/installer/installer.manager.inc#L676-L712). THis would also be helpful in [the issue to add links to the modules page](https://github.com/backdrop/backdrop-issues/issues/5003). 

We can either dump to a file also, which would prevent multiple http requests or can have an endppoint for the installer browser server or project module that returns the list. We're smaller than Drupal for now, but their list on releases alone, without any metadata, is 9MB, hence I suspect the decision to dump to a file 4 times a day instead.

**Comment by stpaultim** [comment]
@BWPanda - I've been playing with your code (see above) tonight and it is really helpful. I'm working on a module to pull in projects and create nodes. It's mostly working, but I'm working on the advanced features now. 

This is both a learning project for me and possibly a proof-of-concept project that I hope might lead to other things. 

I'm inclined to keep this issue open a little longer - until I have my proof of concept done and we are able to talk about whether or not to do something on B.org. I hope this will be soon.

**Comment by stpaultim** [comment]
So here is the work I've been doing on this module. https://github.com/stpaultim/project_feed

The one piece of data that Github does not have is usage data?

Should we provide an API to get usage stats from BackdropCMS.org

**Comment by BWPanda** [comment]
@stpaultim I don't believe you've explained (or I've just missed it) what the overall goal of this project/code is... What are you trying to achieve? Is it something just for your own use/projects, or is it something you intend to include in B.org/core/etc.? If so, will it replace or be integrated into the Project module and the things we already have setup in that regard?

Just interested to hear your goals/intentions here...

**Comment by stpaultim** [comment]
@BWPanda - I've tried to explain it a bit elsewhere, but it's worth repeating here - with a bit more clarity. 

Here are the problems I'm trying to solve.

1) I find it difficult (but not impossible) to find a list of all the projects I maintain. I believe there is a query I can do in GIT to find the projects I have maintainer permisisons on, but it's far from easy or obvious how to do this. 

2) I'd like to it to be eaiser to find projects that are not on BackdropCMS.org. Projects being worked on without releases. I frequently here from members of the community that they are looking for a module and can't find it, even though it exists without an official release. 

3) I'd like to be able to sort modules/projects by tags or categories. The "tags[]" option in projects has not been used enough to be useful and B.org does not provide a way to use it yet anyway. 

If we could add all these features to B.org, I'd be happy. BUT, I'm skeptical that we have the consensus that all of these features are desirable, yet alone any consensus on how they should work or what they should look like. 

My current thought is to set up demo site that provides these options. FIrst as a proof of concept, but potentially as a third party resource for the community. My expectations are this. 

1) I provide a proof of concept site with the features I want. 
2) Once others have had a chance to see the featurs, we can talk about which ones are useful and/or which (if any) of them we would like to implement on B.org.
3) Depending upon the outcome of the decisions in step #2, I can decide if it is worth keeping a third party resource or if enough of the features have been included in B.org that it's not necessary. 

I understand that ideally, we'd just have a single place for all of these features and information. BUT, I also understand that it's much easier to experiment with and test new features on an independent site than it is on B.org. Also, I understand that there might not be consensus on the value of all these features.

Hopefully, this provides some better context for my current experiments.

**Comment by stpaultim** [comment]
One more note. I've been a part of many Open Data projects in the past and recognize that sometimes the availability of good data leads to interesting and creative innovations that one would not have predicted. 

Who knows what kind of innovative things folks might use this data for if it were easily accessible?

**Comment by BWPanda** [comment]
Thanks for the explanation @stpaultim, that all sounds good!

Re. the problems you're solving:
1. I too have this problem. I've kind of solved it by making sure I 'watch'/subscribe to repos that I'm a maintainer of, but all it takes is one time for me to forget to do that, then my workaround doesn't work anymore. And I'm not even sure there's a way I can get a list of repos I'm watching... So your solution would be great to see added somewhere for future use.
2. When looking for a module, I generally only search GitHub since it has *all* projects (as opposed to B.org which lists most, but not all). I just go to https://github.com/orgs/backdrop-contrib/repositories and search there. But I know I'm a developer and GitHub's a developer tool, so providing an easier way to do this for non-developers would be good I guess.
3. If not using tags already in project .info files, where would you get tags from? Or would you be manually adding them?

I wasn't aware of that issue re. problems with GitHub, but understandably it was a long time ago (almost 10 years!), and I'm sure things have changed a lot since then. Be interesting to know @quicksketch's thoughts about it now. But I suspect we won't be moving away from GitHub anytime soon...

**Comment by stpaultim** [comment]
@BWPanda - Funny story: Regarding the issue with Github, that was completely accidental. I did not intent to link to that old issue. I was just intending to reference item `#2` in my own list. But, Github auto linked to that old issue, which by cooincdence was related to Github. 

I now see that I have referenced that issue many times. I'm quite sure that everytime I did, it was an accidental case of typing in `#2` without intending to reference another issue. 

- - - - - - 

In terms of tags, yes I'm considering a manual solution for now. Either doing it myself or by allowing approved users to tag projects. Until I get a better idea. 

I could start submitting PR's to various projects that add tags, but even then we need to implement filters for tags on B.org before they are useful. Maybe, if I prove the usefulness of tags in a third party site it will be motivation to move them forward on B.org and get them into existing projects.

**Comment by stpaultim** [comment]
I believe that @BWPanda is correct. Since most of this data is available through Github, there is probably no need to create an additional API. There is one exception and I've opened a new ticket to address that specific issue:

https://github.com/backdrop-ops/backdropcms.org/issues/912
